memory.go

```diff
@@ -0,0 +1,280 @@
+// This may lead to partially-written data to be observed from user space.
+//
+// On most architectures, the memmove implementation used by Go's copy() will
+// access data in word-sized chunks. If paired with a matching access pattern on
+// the eBPF C side (and if using default memory alignment), accessing shared
+// memory without atomics or other synchronization primitives should be sound
+// for individual values. For accesses beyond a single value, the usual
+// concurrent programming rules apply.
+type Memory struct {
+	b  []byte
+	ro bool
+}
+
+func newMemory(fd, size int, ro bool) (*Memory, error) {
+	// Some architectures need the size to be page-aligned.
+	if internal.Align(size, os.Getpagesize()) != size {
+		return nil, fmt.Errorf("memory: must be a multiple of page size (requested %d bytes)", size)
+	}
+
+	// Allocate a page-aligned span of memory on the Go heap.
+	alloc, err := allocate(size)
+	if err != nil {
+		return nil, fmt.Errorf("allocating memory: %w", err)
+	}
+
+	flags := unix.PROT_READ | unix.PROT_WRITE
+	if ro {
+		flags = unix.PROT_READ
+	}
+
+	// Map the bpf map memory over a page-aligned allocation on the Go heap.
+	if err := mapmap(fd, alloc, size, flags); err != nil {
+		return nil, fmt.Errorf("mapping memory: %w", err)
+	}
+
+	mm := &Memory{
+		unsafe.Slice((*byte)(alloc), size),
+		ro,
+	}
+
+	return mm, nil
+}
+
+// allocate returns a pointer to a page-aligned section of memory on the Go
+// heap, managed by the runtime.
+func allocate(size int) (unsafe.Pointer, error) {
+	// Memory-mapping over a piece of the Go heap is unsafe when the GC can
+	// randomly decide to move objects around, in which case the mapped region
+	// will not move along with it.
+	if heapObjectsCanMove() {
+		return nil, errors.New("this Go runtime has a moving garbage collector")
+	}
+
+	if size == 0 {
+		return nil, errors.New("size must be greater than 0")
+	}
+
+	// Request at least two pages of memory from the runtime to ensure we can
+	// align the requested allocation to a page boundary. This is needed for
+	// MAP_FIXED and makes sure we don't mmap over some other allocation on the Go
+	// heap.
+	size = internal.Align(size+os.Getpagesize(), os.Getpagesize())
+
+	// Allocate a new slice and store a pointer to its backing array.
+	alloc := unsafe.Pointer(unsafe.SliceData(make([]byte, size)))
+
+	// Align the pointer to a page boundary within the allocation. This may
+	// alias the initial pointer if it was already page-aligned.
+	aligned := unsafe.Pointer(internal.Align(uintptr(alloc), uintptr(os.Getpagesize())))
+	runtime.KeepAlive(alloc)
+
+	// Return an aligned pointer into the backing array, losing the original
+	// reference. The runtime.SetFinalizer docs specify that its argument 'must be
+	// a pointer to an object, complit or local var', but this is still somewhat
+	// vague and not enforced by the current implementation.
+	//
+	// Currently, finalizers can be set and triggered from any address within a
```

Comment:
I included this bit since we'll probably switch the whole lib over to `runtime.AddCleanup` as soon as we can use it. I assume the underlying implementation is unlikely to change in the meantime, given the future API will make this explicit.